lib/Provider/MailProvider.php

```diff
@@ -0,0 +1,175 @@
+		return (count($this->listServices($userId)) > 0);
+	}
+
+	/**
+	 * retrieve collection of services for a specific user
+	 *
+	 * @since 4.0.0
+	 *
+	 * @param string $userId system user id
+	 *
+	 * @return array<string,IService> collection of service id and object ['1' => IServiceObject]
+	 */
+	public function listServices(string $userId): array {
+
+		try {
+			// retrieve service(s) details from data store
+			$accounts = $this->accountService->findByUserId($userId);
+		} catch (\Throwable $th) {
+			return [];
+		}
+		// construct temporary collection
+		$services = [];
+		// add services to collection
+		foreach ($accounts as $entry) {
+			// extract values
+			$serviceId = (string)$entry->getId();
+			$label = $entry->getName();
+			$address = new MailAddress($entry->getEmail(), $entry->getName());
+			// add service to collection
+			$services[$serviceId] = new MailService($this->container, $userId, $serviceId, $label, $address);
+		}
+		// return list of services for user
+		return $services;
+
+	}
+
+	/**
+	 * retrieve a service with a specific id
+	 *
+	 * @since 4.0.0
+	 *
+	 * @param string $userId system user id
+	 * @param string $serviceId mail account id
+	 *
+	 * @return IService|null returns service object or null if none found
+	 *
+	 */
+	public function findServiceById(string $userId, string $serviceId): IService|null {
+		
+		// determine if a valid user and service id was submitted
+		if (empty($userId) && !ctype_digit($serviceId)) {
+			return null;
+		}
+
+		try {
+			// retrieve service details from data store
+			$account = $this->accountService->find($userId, (int)$serviceId);
+		} catch(\Throwable $th) {
+			return null;
+		}
+
+		// extract values
+		$serviceId = (string)$account->getId();
+		$label = $account->getName();
+		$address = new MailAddress($account->getEmail(), $account->getName());
+		// return mail service object
+		return new MailService($this->container, $userId, $serviceId, $label, $address);
+
+	}
+
+	/**
+	 * retrieve a service for a specific mail address
+	 *
+	 * @since 4.0.0
+	 *
+	 * @param string $userId system user id
+	 * @param string $address mail address (e.g. test@example.com)
+	 *
+	 * @return IService|null returns service object or null if none found
+	 */
+	public function findServiceByAddress(string $userId, string $address): IService|null {
+
+		try {
+			// retrieve service details from data store
+			$accounts = $this->accountService->findByUserIdAndAddress($userId, $address);
+		} catch(\Throwable $th) {
+			return null;
+		}
+		// evaluate if service details where found
+		if (isset($accounts[0])) {
+			// extract values
+			$serviceId = (string)$accounts[0]->getId();
+			$serviceLabel = $accounts[0]->getName();
+			$serviceAddress = new MailAddress($accounts[0]->getEmail(), $accounts[0]->getName());
+			// return mail service object
+			return new MailService($this->container, $userId, $serviceId, $serviceLabel, $serviceAddress);
```

Comment:
We have the code to create a mail service instance three times (findServiceByAddress, findServiceById, listServices). Please add a factory method MailService::createFromAccount($this->container, $userId, $account) to avoid the duplicate code.